### PR TITLE
#13998 DistSQL to support `SHOW SHARDING KEY GENERATORS`

### LINKED ISSUE #13998: Add syntax for `SHOW SHARDING KEY GENERATORS`
For details, please refer to https://github.com/apache/shardingsphere/issues/13994

The ANTLR definition syntax is as follows:

```sql
SHOW SHARDING KEY GENERATORS (FROM schemaName)?
```

Examples are as follows:

```sql
> SHOW SHARDING KEY GENERATORS;
+-------------------------+--------+-------------------+
| name                    | type   | props             |
+-------------------------+----------------------------+
| uuid_key_generator      | uuid   | "worker-id"=123   |
+-------------------------+--------+-------------------+
```

### PR DESCRIPTION
**Pull request by CrackerSuperior**
Fixes #13998 .

Changes proposed in this pull request:

Add syntax for `SHOW SHARDING KEY GENERATORS`
Add corresponding test

### CODE REVIEW

**Comment by lanchengx** [comment]
Hi @CrackerSuperior , 
I found that CI has some errors. You can execute `mvn clean install` locally after the development is completed to ensure that the compilation passes once to reduce CI errors.

In addition, you can refer to the development specification https://shardingsphere.apache.org/community/cn/contribute/code-conduct/

**Comment by CrackerSuperior** [comment]
image
How should this be obtained? @lanchengx

**Comment by lanchengx** [comment]
Copy from the header of an existing file.

**Comment by CrackerSuperior** [comment]
> Hi @CrackerSuperior , I found that CI has some errors. You can execute `mvn clean install` locally after the development is completed to ensure that the compilation passes once to reduce CI errors.
> 
> In addition, you can refer to the development specification https://shardingsphere.apache.org/community/cn/contribute/code-conduct/

image
image

If you don't skip the javadoc, you will get an error, what should I do?

**Comment by lanchengx** [comment]
I did not find the `cmd.exe` file in the corresponding directory, maybe you can skip the javadoc locally, CI will verify it

**Comment by CrackerSuperior** [comment]
> I did not find the `cmd.exe` file in the corresponding directory, maybe you can skip the javadoc locally, CI will verify it


Local inspection passed

**Comment by lanchengx** [comment]
Synchronized to the remote branch.

**Comment by CrackerSuperior** [comment]
> Synchronized to the remote branch.


this?

**Comment by lanchengx** [comment]
Yes, wait for CI and review

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingKeyGeneratorsQueryResultSet.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    private static final String TYPE = "type";
+
+    private static final String PROPS = "props";
+
+    private Iterator<Entry<String, ShardingSphereAlgorithmConfiguration>> data = Collections.emptyIterator();
+
+    @Override
+    public void init(final ShardingSphereMetaData metaData, final SQLStatement sqlStatement) {
+
+        metaData.getRuleMetaData().getConfigurations().stream()
```
Comment (lanchengx) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingKeyGeneratorsQueryResultSet.java:50]: Redundant blank line

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingKeyGeneratorsQueryResultSet.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+    private static final String TYPE = "type";
+
+    private static final String PROPS = "props";
+
+    private Iterator<Entry<String, ShardingSphereAlgorithmConfiguration>> data = Collections.emptyIterator();
+
+    @Override
+    public void init(final ShardingSphereMetaData metaData, final SQLStatement sqlStatement) {
+
```
Comment (lanchengx) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingKeyGeneratorsQueryResultSet.java:49]: Redundant blank line

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/test/java/org/apache/shardingsphere/sharding/distsql/query/ShardingKeyGeneratorsQueryResultSetTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    private Collection<RuleConfiguration> createRuleConfigurations() {
+        ShardingRuleConfiguration result = new ShardingRuleConfiguration();
+        Properties props = new Properties();
+        props.put("work-id", 123);
+        result.getKeyGenerators().put("snowflake", new ShardingSphereAlgorithmConfiguration("SNOWFLAKE", props));
+        return Collections.singleton(result);
+    }
+
```
Comment (lanchengx) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/test/java/org/apache/shardingsphere/sharding/distsql/query/ShardingKeyGeneratorsQueryResultSetTest.java:62]: Redundant blank line

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingKeyGeneratorsStatementAssert.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class ShowShardingKeyGeneratorsStatementAssert {
+
+    public static void assertIs(final SQLCaseAssertContext assertContext, final ShowShardingKeyGeneratorsStatement actual,
+                                final ShowShardingKeyGeneratorsStatementTestCase expected) {
+        assertTrue(assertContext.getText("Actual schema should exist."), actual.getSchema().isPresent());
+        SchemaAssert.assertIs(assertContext, actual.getSchema().get(), expected.getSchema());
+    }
+
```
Comment (lanchengx) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingKeyGeneratorsStatementAssert.java:37]: Redundant blank line

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/distsql/rql/ShowShardingKeyGeneratorsStatementTestCase.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Show sharding key generators statement test case.
+ */
+@Getter
+@Setter
+public final class ShowShardingKeyGeneratorsStatementTestCase extends SQLParserTestCase {
+
+    @XmlElement
+    private ExpectedSchema schema;
+
```
Comment (lanchengx) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/statement/distsql/rql/ShowShardingKeyGeneratorsStatementTestCase.java:36]: Redundant blank line

File: shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/resources/conf/server.yaml [hunk 6]
```diff
@@ -84,4 +84,4 @@
 #    # if client connections are more than proxy-frontend-netty-executor-size, especially executing slow SQL.
 #  proxy-backend-executor-suitable: OLAP
 #  proxy-frontend-max-connections: 0 # Less than or equal to 0 means no limitation.
-#  sql-federation-enabled: false
+#  sql-federation-enabled: false
```
Comment (lanchengx) [inline shardingsphere-proxy/shardingsphere-proxy-bootstrap/src/main/resources/conf/server.yaml:87]: The file needs a blank line at the end

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingKeyGeneratorsQueryResultSet.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        metaData.getRuleMetaData().getConfigurations().stream()
+                .filter(each -> each instanceof ShardingRuleConfiguration)
+                .map(each -> (ShardingRuleConfiguration) each)
+                .forEach(each -> data = each.getKeyGenerators().entrySet().iterator());
```
Comment (lanchengx) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/query/ShardingKeyGeneratorsQueryResultSet.java:53]: Consider using the  `ShardingSphereRuleMetaData#findRuleConfiguration` method

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/ShardingDistSQLStatement.g4 [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    | showShardingHintStatus
+    | clearShardingHint
+    | createShardingAlgorithm
+    | alterShardingAlgorithm
+    | showShardingTableNodes
+    | createShardingKeyGenerator
+    | alterShardingKeyGenerator
+    | showShardingKeyGenerators
+    ) SEMI?
+    ;
```
Comment (lanchengx) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/ShardingDistSQLStatement.g4:50]: Shouldn't be added, it's the same as
`src/main/antlr4/org/apache/shardingsphere/distsql/parser/autogen/ShardingDistSQLStatement.g4` same

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/java/org/apache/shardingsphere/sharding/distsql/parser/core/ShardingDistSQLStatementVisitor.java [hunk 9]
```diff
@@ -365,4 +367,11 @@ public ASTNode visitAlterShardingKeyGenerator(final AlterShardingKeyGeneratorCon
     private ShardingKeyGeneratorSegment buildShardingKeyGeneratorSegment(final KeyGeneratorDefinationContext ctx) {
         return new ShardingKeyGeneratorSegment(getIdentifierValue(ctx.keyGeneratorName()), (AlgorithmSegment) visitAlgorithmDefinition(ctx.algorithmDefinition()));
     }
+
+    @Override
+    public ASTNode visitShowShardingKeyGenerators(final ShardingDistSQLStatementParser.ShowShardingKeyGeneratorsContext ctx) {
+        return new ShowShardingKeyGeneratorsStatement(Objects.nonNull(ctx.schemaName())
+                ? (SchemaSegment) visit(ctx.schemaName()) : null);
+    }
+
```
Comment (lanchengx) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/java/org/apache/shardingsphere/sharding/distsql/parser/core/ShardingDistSQLStatementVisitor.java:376]: Redundant blank line

File: shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingKeyGeneratorsStatementAssert.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.shardingsphere.sharding.distsql.parser.statement.ShowShardingKeyGeneratorsStatement;
+import org.apache.shardingsphere.test.sql.parser.parameterized.asserts.SQLCaseAssertContext;
+import org.apache.shardingsphere.test.sql.parser.parameterized.asserts.segment.distsql.SchemaAssert;
+import org.apache.shardingsphere.test.sql.parser.parameterized.jaxb.cases.domain.statement.distsql.rql.ShowShardingKeyGeneratorsStatementTestCase;
+
+import static org.junit.Assert.assertTrue;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class ShowShardingKeyGeneratorsStatementAssert {
+
```
Comment (lanchengx) [inline shardingsphere-test/shardingsphere-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/asserts/statement/distsql/rql/impl/rule/ShowShardingKeyGeneratorsStatementAssert.java:31]: Public methods require javadoc

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/Keyword.g4 [hunk 11]
```diff
@@ -178,3 +178,7 @@ KEY
 GENERATOR
     : G E N E R A T O R
     ;
+
+GENERATORS
+    : GENERATOR S
```
Comment (RaigorJiang) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/Keyword.g4:183]: keyword should not quote others, it is recommended to change to `G E N E R A T O R S`

**Comment by RaigorJiang** [comment]
Hi @CrackerSuperior,
Please resolve the conflict, maybe just merge from master and push again.

**Review by RaigorJiang — APPROVED** [review]

**Comment by RaigorJiang** [comment]
@CrackerSuperior 
Merged, thanks for your great work 👍

**Comment by CrackerSuperior** [comment]
> @CrackerSuperior Merged, thanks for your great work 👍

:) , Hey-hey, thanks

**Pull request merged**